lib/pure/algorithm.nim

```diff
@@ -556,6 +558,76 @@ proc isSorted*[T](a: openArray[T], order = SortOrder.Ascending): bool =
     assert isSorted(e) == false
   isSorted(a, system.cmp[T], order)
 
+proc merge*[T](
+  x, y: openArray[T], cmp: proc(x, y: T
+): int {.closure.}, order = SortOrder.Ascending): seq[T] {.since: (1, 5, 1).} =
+  ## Merges two sorted `openArray`. All of inputs are assumed to be sorted.
+  ## If you do not wish to provide your own ``cmp``,
+  ## you may use `system.cmp` or instead call the overloaded
+  ## version of `merge`, which uses `system.cmp`.
+  ##
+  ## **See also:**
+  ## * `merge proc<#merge,openArray[T],openArray[T]>`_
+  runnableExamples:
+    let x = @[1, 3, 6]
+    let y = @[2, 3, 4]
+
+    let res = x.merge(y, system.cmp[int])
+    assert res.isSorted
+    assert res == @[1, 2, 3, 3, 4, 6]
+
+  let
+    sizeX = x.len
+    sizeY = y.len
+
+  result = newSeq[T](sizeX + sizeY)
+
+  var
+    ix = 0
+    iy = 0
+    i = 0
+
+  while true:
+    if ix == sizeX:
+      while iy < sizeY:
+        result[i] = y[iy]
+        inc i
+        inc iy
+      return
+
+    if iy == sizeY:
+      while ix < sizeX:
+        result[i] = x[ix]
+        inc i
+        inc ix
+      return
+
+    let itemX = x[ix]
+    let itemY = y[iy]
+
+    if cmp(itemX, itemY) * order > 0:
+      result[i] = itemY
+      inc iy
+    else:
+      result[i] = itemX
+      inc ix
+
+    inc i
+
+proc merge*[T](x, y: openArray[T], order = SortOrder.Ascending): seq[T] {.since: (1, 5, 1).} =
```

Comment:
i know other procs don't do this but can't we use an optional param instead?

Reply:
Yeah, I don't know why other procs do that. I will use optional param instaed.

Reply:
yeah, but it is a bit inconsistent with `sort` when calling it. 
```nim
merge(a, b, order=order)
```

```nim
sort(a, b, order)
```

Reply:
order is redundant though since you can encode this in cmp (yes, I know it's used in other procs)

Reply:
I think it's better to be consistent with existing procs since we could clean them all in the future. Anyway not too bad.

Reply:
but `order` is expensive, see https://github.com/timotheecour/Nim/issues/400

Reply:
Is that big difference? And This is a bit convenient to users?

Reply:
The difference is significant enough; I'd rather not pay for this "convenience", otherwise it just invites users to re-implement this, which isn't practical because other algorithms are gonna depend on algorithm.merge. The old design (cmp+SortOrder.Ascending) is bad IMO and should'n't be copied over to new procs.

Reply:
I see. I will clean them later.

Reply:
What I should do: add `SortOrder.Ascending` to cmp parameter lists?

Reply:
The convenience is nice, we should fix the codegen for it if it slows things down. There is no reason it should though.